### DISCUSSION TITLE
torch-cluster issue: No CUDA runtime is found

### DISCUSSION THREAD

**Discussion by YoavShamir5** [pyg-team/pytorch_geometric#7424] [Q&A] [answered]
Hi, I generated a new conda env on windows with the following commands:

conda create -n envn python=3.10.9
conda install python=3.10.9 pytorch torchvision pytorch-cuda=11.7 -c pytorch -c nvidia
conda install cuda=11.7 -c nvidia 
conda install pyg -c pyg

Then, I tried to install torch-cluster:
pip install torch_cluster -f https://data.pyg.org/whl/torch-2.0.1+cu117.html --no-cache-dir

But got an error message. Any ideas how could this be resolved? Please see error message and output of conda list below. Thanks!

error:

(envn) C:\Users\SYOAV>C:\Users\SYOAV\anaconda3\envs\envn\Scripts\pip install torch-cluster -f https://data.pyg.org/whl/torch-2.0.1+cu117.html --no-cache-dir 
Looking in links: https://data.pyg.org/whl/torch-2.0.1+cu117.html
Collecting torch-cluster
  Downloading torch_cluster-1.6.1.tar.gz (53 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.8/53.8 kB 463.1 kB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
Requirement already satisfied: scipy in c:\users\syoav\anaconda3\envs\envn\lib\site-packages (from torch-cluster) (1.10.1)
Requirement already satisfied: numpy<1.27.0,>=1.19.5 in c:\users\syoav\anaconda3\envs\envn\lib\site-packages (from scipy->torch-cluster) (1.24.3)
Building wheels for collected packages: torch-cluster
  Building wheel for torch-cluster (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [32 lines of output]
      No CUDA runtime is found, using CUDA_HOME='C:\Users\SYOAV\anaconda3\envs\envn'
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-cpython-310
      creating build\lib.win-amd64-cpython-310\torch_cluster
      copying torch_cluster\fps.py -> build\lib.win-amd64-cpython-310\torch_cluster
      copying torch_cluster\graclus.py -> build\lib.win-amd64-cpython-310\torch_cluster
      copying torch_cluster\grid.py -> build\lib.win-amd64-cpython-310\torch_cluster
      copying torch_cluster\knn.py -> build\lib.win-amd64-cpython-310\torch_cluster
      copying torch_cluster\nearest.py -> build\lib.win-amd64-cpython-310\torch_cluster
      copying torch_cluster\radius.py -> build\lib.win-amd64-cpython-310\torch_cluster
      copying torch_cluster\rw.py -> build\lib.win-amd64-cpython-310\torch_cluster
      copying torch_cluster\sampler.py -> build\lib.win-amd64-cpython-310\torch_cluster
      copying torch_cluster\testing.py -> build\lib.win-amd64-cpython-310\torch_cluster
      copying torch_cluster\__init__.py -> build\lib.win-amd64-cpython-310\torch_cluster
      running egg_info
      writing torch_cluster.egg-info\PKG-INFO
      writing dependency_links to torch_cluster.egg-info\dependency_links.txt
      writing requirements to torch_cluster.egg-info\requires.txt
      writing top-level names to torch_cluster.egg-info\top_level.txt
      reading manifest file 'torch_cluster.egg-info\SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      warning: no previously-included files matching '*' found under directory 'test'
      adding license file 'LICENSE'
      writing manifest file 'torch_cluster.egg-info\SOURCES.txt'
      running build_ext
      C:\Users\SYOAV\anaconda3\envs\envn\lib\site-packages\torch\utils\cpp_extension.py:359: UserWarning: Error checking compiler version for cl: [WinError 2] The system cannot find the file specified
        warnings.warn(f'Error checking compiler version for {compiler}: {error}')
      building 'torch_cluster._fps_cpu' extension
      error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for torch-cluster
  Running setup.py clean for torch-cluster
Failed to build torch-cluster
ERROR: Could not build wheels for torch-cluster, which is required to install pyproject.toml-based projects

conda list:
(envn) C:\Users\SYOAV>conda list
packages in environment at C:\Users\SYOAV\anaconda3\envs\envn:        

Name                    Version                   Build  Channel      
blas                      2.116                       mkl    conda-forge
blas-devel                3.9.0              16_win64_mkl    conda-forge
brotli                    1.0.9                hcfcfb64_8    conda-forge
brotli-bin                1.0.9                hcfcfb64_8    conda-forge
bzip2                     1.0.8                h8ffe710_4    conda-forge
ca-certificates           2023.5.7             h56e8100_0    conda-forge
certifi                   2023.5.7           pyhd8ed1ab_0    conda-forge
charset-normalizer        3.1.0              pyhd8ed1ab_0    conda-forge
colorama                  0.4.6              pyhd8ed1ab_0    conda-forge
cuda                      11.7.1                        0    nvidia     
cuda-cccl                 12.1.109                      0    nvidia     
cuda-command-line-tools   11.7.1                        0    nvidia     
cuda-compiler             12.1.1                        0    nvidia     
cuda-cudart               11.7.99                       0    nvidia     
cuda-cudart-dev           11.7.99                       0    nvidia     
cuda-cuobjdump            12.1.111                      0    nvidia     
cuda-cupti                11.7.101                      0    nvidia     
cuda-cuxxfilt             12.1.105                      0    nvidia     
cuda-demo-suite           12.1.105                      0    nvidia     
cuda-documentation        12.1.105                      0    nvidia     
cuda-libraries            11.7.1                        0    nvidia     
cuda-libraries-dev        11.7.1                        0    nvidia     
cuda-memcheck             11.8.86                       0    nvidia     
cuda-nsight-compute       12.1.1                        0    nvidia     
cuda-nvcc                 12.1.105                      0    nvidia     
cuda-nvdisasm             12.1.105                      0    nvidia     
cuda-nvml-dev             12.1.105                      0    nvidia     
cuda-nvprof               12.1.105                      0    nvidia     
cuda-nvprune              12.1.105                      0    nvidia     
cuda-nvrtc                11.7.99                       0    nvidia     
cuda-nvrtc-dev            11.7.99                       0    nvidia     
cuda-nvtx                 11.7.91                       0    nvidia     
cuda-nvvp                 12.1.105                      0    nvidia     
cuda-runtime              11.7.1                        0    nvidia     
cuda-sanitizer-api        12.1.105                      0    nvidia     
cuda-toolkit              11.7.1                        0    nvidia     
cuda-tools                11.7.1                        0    nvidia     
cuda-visual-tools         11.7.1                        0    nvidia     
filelock                  3.12.0             pyhd8ed1ab_0    conda-forge
freetype                  2.12.1               h546665d_1    conda-forge
idna                      3.4                pyhd8ed1ab_0    conda-forge
intel-openmp              2023.1.0         h57928b3_46319    conda-forge
jinja2                    3.1.2              pyhd8ed1ab_1    conda-forge
joblib                    1.2.0              pyhd8ed1ab_0    conda-forge
jpeg                      9e                   hcfcfb64_3    conda-forge
lcms2                     2.15                 ha5c8aab_0    conda-forge
lerc                      4.0.0                h63175ca_0    conda-forge
libblas                   3.9.0              16_win64_mkl    conda-forge
libbrotlicommon           1.0.9                hcfcfb64_8    conda-forge
libbrotlidec              1.0.9                hcfcfb64_8    conda-forge
libbrotlienc              1.0.9                hcfcfb64_8    conda-forge
libcblas                  3.9.0              16_win64_mkl    conda-forge
libcublas                 11.10.3.66                    0    nvidia
libcublas-dev             11.10.3.66                    0    nvidia
libcufft                  10.7.2.124                    0    nvidia
libcufft-dev              10.7.2.124                    0    nvidia
libcurand                 10.3.2.106                    0    nvidia
libcurand-dev             10.3.2.106                    0    nvidia
libcusolver               11.4.0.1                      0    nvidia
libcusolver-dev           11.4.0.1                      0    nvidia
libcusparse               11.7.4.91                     0    nvidia
libcusparse-dev           11.7.4.91                     0    nvidia
libdeflate                1.17                 hcfcfb64_0    conda-forge
libffi                    3.4.2                h8ffe710_5    conda-forge
libhwloc                  2.9.1                h51c2c0f_0    conda-forge
libiconv                  1.17                 h8ffe710_0    conda-forge
liblapack                 3.9.0              16_win64_mkl    conda-forge
liblapacke                3.9.0              16_win64_mkl    conda-forge
libnpp                    11.7.4.75                     0    nvidia
libnpp-dev                11.7.4.75                     0    nvidia
libnvjpeg                 11.8.0.2                      0    nvidia
libnvjpeg-dev             11.8.0.2                      0    nvidia
libpng                    1.6.39               h19919ed_0    conda-forge
libsqlite                 3.42.0               hcfcfb64_0    conda-forge
libtiff                   4.5.0                hf8721a0_2    conda-forge
libuv                     1.44.2               h8ffe710_0    conda-forge
libwebp-base              1.3.0                hcfcfb64_0    conda-forge
libxcb                    1.13              hcd874cb_1004    conda-forge
libxml2                   2.10.4               hc3477c8_0    conda-forge
libzlib                   1.2.13               hcfcfb64_4    conda-forge
m2w64-gcc-libgfortran     5.3.0                         6    conda-forge
m2w64-gcc-libs            5.3.0                         7    conda-forge
m2w64-gcc-libs-core       5.3.0                         7    conda-forge
m2w64-gmp                 6.1.0                         2    conda-forge
m2w64-libwinpthread-git   5.0.0.4634.697f757               2    conda-forge
markupsafe                2.1.2           py310h8d17308_0    conda-forge
mkl                       2022.1.0           h6a75c08_874    conda-forge
mkl-devel                 2022.1.0           h57928b3_875    conda-forge
mkl-include               2022.1.0           h6a75c08_874    conda-forge
mpmath                    1.3.0              pyhd8ed1ab_0    conda-forge
msys2-conda-epoch         20160418                      1    conda-forge
networkx                  3.1                pyhd8ed1ab_0    conda-forge
nsight-compute            2023.1.1.4                    0    nvidia
numpy                     1.24.3          py310hd02465a_0    conda-forge
openjpeg                  2.5.0                ha2aaf27_2    conda-forge
openssl                   3.1.0                hcfcfb64_3    conda-forge
packaging                 23.1               pyhd8ed1ab_0    conda-forge
pillow                    9.4.0           py310hdbb7713_1    conda-forge
pip                       23.1.2             pyhd8ed1ab_0    conda-forge
platformdirs              3.5.1              pyhd8ed1ab_0    conda-forge
pooch                     1.7.0              pyha770c72_3    conda-forge
psutil                    5.9.5           py310h8d17308_0    conda-forge
pthread-stubs             0.4               hcd874cb_1001    conda-forge
pthreads-win32            2.9.1                hfa6e2cd_3    conda-forge
pyg                       2.3.0           py310_torch_2.0.0_cu117    pyg
pyparsing                 3.0.9              pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1              pyh0701188_6    conda-forge
python                    3.10.9          h4de0772_0_cpython    conda-forge
python_abi                3.10                    3_cp310    conda-forge
pytorch                   2.0.1           py3.10_cuda11.7_cudnn8_0    pytorch
pytorch-cuda              11.7                 h16d0643_5    pytorch
pytorch-mutex             1.0                        cuda    pytorch
requests                  2.31.0             pyhd8ed1ab_0    conda-forge
scikit-learn              1.2.2           py310hd266714_1    conda-forge
scipy                     1.10.1          py310h578b7cb_3    conda-forge
setuptools                67.7.2             pyhd8ed1ab_0    conda-forge
sqlite                    3.42.0               hcfcfb64_0    conda-forge
sympy                     1.12               pyh04b8f61_3    conda-forge
tbb                       2021.9.0             h91493d7_0    conda-forge
threadpoolctl             3.1.0              pyh8a188c0_0    conda-forge
tk                        8.6.12               h8ffe710_0    conda-forge
torchvision               0.15.2                   pypi_0    pypi
tqdm                      4.65.0             pyhd8ed1ab_1    conda-forge
typing-extensions         4.6.0                hd8ed1ab_0    conda-forge
typing_extensions         4.6.0              pyha770c72_0    conda-forge
tzdata                    2023c                h71feb2d_0    conda-forge
ucrt                      10.0.22621.0         h57928b3_0    conda-forge
urllib3                   2.0.2              pyhd8ed1ab_0    conda-forge
vc                        14.3                hb25d44b_16    conda-forge
vc14_runtime              14.34.31931         h5081d32_16    conda-forge
vs2015_runtime            14.34.31931         hed1258a_16    conda-forge
wheel                     0.40.0             pyhd8ed1ab_0    conda-forge
win_inet_pton             1.1.0              pyhd8ed1ab_6    conda-forge
xorg-libxau               1.0.11               hcd874cb_0    conda-forge
xorg-libxdmcp             1.1.3                hcd874cb_0    conda-forge
xz                        5.4.2                h8cc25b3_0
zlib                      1.2.13               hcfcfb64_4    conda-forge
zstd                      1.5.2                h12be248_6    conda-forge

**Comment by wsad1** [comment]
Windows binaries are not available for conda yet (see [here](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/kge/rotate.html#RotatE)). Try using pip for your installation.

**Reply by YoavShamir5** [reply]
Hi, thanks for the quick reply - do you mean using pip to install torch_geometric? I installed with conda and that seems to work, only torch_cluster, torch_scatter etc. were resulting in an error.

**Reply by YoavShamir5** [reply]
all good now - thanks for the quick reply!

**Comment by rusty1s** [comment, marked as answer]
Also try
```
pip install torch_cluster -f https://data.pyg.org/whl/torch-2.0.0+cu117.html --no-cache-dir
```
It looks like it currently tries to build from source (due to the missing `torch-2.0.1+cu117.html` page on our end).

**Reply by YoavShamir5** [reply]
That solved it - thanks for helping me out so quickly!
